File: zh_CN/commu_a_c.py
import math
import sys

sgn = lambda n: 1 if n > 0 else -1 if n < 0 else 0
fndeg = lambda d, m, s: d + sgn(d) * (m / 60 + s/ 3600)
fnacs = lambda x: math.pi / 2 - 2 * math.atan(x / (1 + math.sqrt(1 - x * x)))
    
def calc(tlongd, tlatd, rlongd, rlatd):
    ro = 6370

    tlong = math.radians(tlongd)
    tlat = math.radians(tlatd)
    rlong = math.radians(rlongd)
    rlat = math.radians(rlatd)

    if tlong < 0:
        tlong = math.pi * 2 + tlong
    if rlong < 0:
        rlong = math.pi * 2 + rlong
    
    dlong = tlong - rlong
    if abs(dlong) > math.pi:
        dlong = dlong - math.pi * 2 * sgn(dlong)
    
    x = math.sin(tlat) * math.sin(rlat) + math.cos(tlat) * math.cos(rlat) * math.cos(dlong)
    gcd = fnacs(x)

    if gcd < 0.0000001:
        gcd = 0.0000001
    
    if math.cos(tlat) - 0.0000001 <= 0:
        if tlat >= 0:
            btr = 0
        else:
            btr = math.pi
        
        if math.cos(rlat) - 0.0000001 > 0:
            x = (math.sin(tlat) - math.sin(rlat) * math.cos(gcd)) / (math.cos(rlat) * math.sin(gcd))
            brt = fnacs(x)
            if dlong < 0:
                brt = math.pi * 2 - brt
            gcdkm = gcd * ro
            btrd = math.degrees(btr)
            brtd = math.degrees(brt)
        else:
            if rlat >= 0:
                brt = 0
            else:
                brt = math.pi
            gcdkm = gcd * ro
            btrd = math.degrees(btr)
            brtd = math.degrees(brt)
    else:
        x = (math.sin(rlat) - math.sin(tlat) * math.cos(gcd)) / (math.cos(tlat) * math.sin(gcd))
        btr = fnacs(x)

        if dlong > 0:
            btr = math.pi * 2 - btr
            
        if math.cos(rlat) - 0.0000001 > 0:
            x = (math.sin(tlat) - math.sin(rlat) * math.cos(gcd)) / (math.cos(rlat) * math.sin(gcd))
            brt = fnacs(x)
            if dlong < 0:
                brt = math.pi * 2 - brt
            gcdkm = gcd * ro
            btrd = math.degrees(btr)
            brtd = math.degrees(brt)
        else:
            if rlat >= 0:
                brt = 0
            else:
                brt = math.pi
            gcdkm = gcd * ro
            btrd = math.degrees(btr)
            brtd = math.degrees(brt)
    
    return gcdkm, btrd, brtd


def main():
    print("《计算通信方位角和大圆距离程序》")

    while True:
        tmp = input("发信点经度（度，分，秒）=")
        try:
            tmp = [int(t) for t in tmp.split(",")]
            tlongd = fndeg(tmp[0], tmp[1], tmp[2])
        except Exception as e:
            print("错误：", e)
        else:
            break
    while True:
        tmp = input("发信点纬度（度，分，秒）=")
        try:
            tmp = [int(t) for t in tmp.split(",")]
            tlatd = fndeg(tmp[0], tmp[1], tmp[2])
        except Exception as e:
            print("错误：", e)
        else:
            break
    while True:
        tmp = input("收信点经度（度，分，秒）=")
        try:
            tmp = [int(t) for t in tmp.split(",")]
            rlongd = fndeg(tmp[0], tmp[1], tmp[2])
        except Exception as e:
            print("错误：", e)
        else:
            break
    while True:
        tmp = input("收信点纬度（度，分，秒）=")
        try:
            tmp = [int(t) for t in tmp.split(",")]
            rlatd = fndeg(tmp[0], tmp[1], tmp[2])
        except Exception as e:
            print("错误：", e)
        else:
            break

    gcdkm, btrd, brtd = calc(tlongd, tlatd, rlongd, rlatd)

    print("大圆距离为", gcdkm, "(km)")
    d = int(btrd)
    m = int((btrd - d) * 60)
    s = (btrd - d - m / 60) * 3600
    print("发信点对收信点的方位角为", d, "度", m, "分", s, "秒")
    d = int(brtd)
    m = int((brtd - d) * 60)
    s = (brtd - d - m / 60) * 3600
    print("收信点对发信点的方位角为", d, "度", m, "分", s, "秒")

if __name__ == "__main__":
    while True:
        main()
        while True:
            a = input("您是否对其他收信点继续计算？(Y/N)")
            if a.lower() not in ("y", "n"):
                continue
            if a.lower() == "y":
                break
            else:
                sys.exit(0)

File: zh_CN/commu_a_c.js
const readline = require("readline");
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function sgn(n) {
    if (n > 0) {
        return 1;
    } else if (n < 0) {
        return -1;
    } else {
        return 0;
    }
}

function fndeg(d, m, s) {
    return d + sgn(d) * (m / 60 + s/ 3600);
}

function fnacs(x) {
    return Math.PI / 2 - 2 * Math.atan(x / (1 + Math.sqrt(1 - x * x)));
}

function calc(tlongd, tlatd, rlongd, rlatd) {
    let ro = 6370;

    let gcdkm;
    let btrd;
    let brtd;

    let r2d = 180 / Math.PI;
    let d2r = Math.PI / 180;

    let tlong = tlongd * d2r;
    let tlat = tlatd * d2r;
    let rlong = rlongd * d2r;
    let rlat = rlatd * d2r;

    if (tlong < 0) {
        tlong = Math.PI * 2 + tlong;
    }
    if (rlong < 0) {
        rlong = Math.PI * 2 + rlong;
    }
    
    let dlong = tlong - rlong;
    if (Math.abs(dlong) > Math.PI) {
        dlong = dlong - Math.PI * 2 * sgn(dlong);
    }
    
    let x = Math.sin(tlat) * Math.sin(rlat) + Math.cos(tlat) * Math.cos(rlat) * Math.cos(dlong);
    let gcd = fnacs(x);

    if (gcd < 0.0000001) {
        gcd = 0.0000001;
    }
    
    if (Math.cos(tlat) - 0.0000001 <= 0) {
        let btr;
        if (tlat >= 0) {
            btr = 0;
        } else {
            btr = Math.PI;
        }
        
        if (Math.cos(rlat) - 0.0000001 > 0) {
            x = (Math.sin(tlat) - Math.sin(rlat) * Math.cos(gcd)) / (Math.cos(rlat) * Math.sin(gcd));
            let brt = fnacs(x);
            if (dlong < 0) {
                brt = Math.PI * 2 - brt;
            }
            gcdkm = gcd * ro;
            btrd = btr * r2d;
            brtd = brt * r2d;
        } else {
            let brt;
            if (rlat >= 0) {
                brt = 0;
            } else {
                brt = Math.PI;
            }
            gcdkm = gcd * ro;
            btrd = btr * r2d;
            brtd = brt * r2d;
        }
    } else {
        x = (Math.sin(rlat) - Math.sin(tlat) * Math.cos(gcd)) / (Math.cos(tlat) * Math.sin(gcd));
        let btr = fnacs(x);

        if (dlong > 0) {
            btr = Math.PI * 2 - btr;
        }
            
        if (Math.cos(rlat) - 0.0000001 > 0) {
            x = (Math.sin(tlat) - Math.sin(rlat) * Math.cos(gcd)) / (Math.cos(rlat) * Math.sin(gcd));
            let brt = fnacs(x);
            if (dlong < 0) {
                brt = Math.PI * 2 - brt;
            }
            gcdkm = gcd * ro;
            btrd = btr * r2d;
            brtd = brt * r2d;
        } else {
            let brt;
            if (rlat >= 0) {
                brt = 0;
            } else {
                brt = Math.PI;
            }
            gcdkm = gcd * ro;
            btrd = btr * r2d;
            brtd = brt * r2d;
        }
    }
    
    return [gcdkm, btrd, brtd];
}

function main() {
    console.log("《计算通信方位角和大圆距离程序》")

    let gcdkm;
    let btrd;
    let brtd;

    rl.question("发信点经度（度，分，秒）=", function(tmp) {
        tmp = tmp.toString().split(",")
        let tmp2 = [];
        for (let i=0;i<tmp.length;i++) {
            tmp2.push(parseInt(tmp[i]))
        }
        let tlongd = fndeg(tmp2[0], tmp2[1], tmp2[2]);
        rl.question("发信点纬度（度，分，秒）=", function(tmp3) {
            tmp3 = tmp3.toString().split(",")
            let tmp4 = [];
            for (let i=0;i<tmp3.length;i++) {
                tmp4.push(parseInt(tmp3[i]))
            }
            let tlatd = fndeg(tmp4[0], tmp4[1], tmp4[2]);
            rl.question("收信点经度（度，分，秒）=", function(tmp5) {
                tmp5 = tmp5.toString().split(",")
                let tmp6 = [];
                for (let i=0;i<tmp5.length;i++) {
                    tmp6.push(parseInt(tmp5[i]))
                }
                let rlongd = fndeg(tmp6[0], tmp6[1], tmp6[2]);
                rl.question("收信点纬度（度，分，秒）=", function(tmp7) {
                    tmp7 = tmp7.toString().split(",")
                    let tmp8 = [];
                    for (let i=0;i<tmp7.length;i++) {
                        tmp8.push(parseInt(tmp7[i]))
                    }
                    let rlatd = fndeg(tmp8[0], tmp8[1], tmp8[2]);
        
                    let temp = calc(tlongd, tlatd, rlongd, rlatd);
                    let gcdkm = temp[0];
                    let btrd = temp[1];
                    let brtd = temp[2];

                    console.log("大圆距离为" + gcdkm + "(km)");
                    let dt = Math.trunc(btrd);
                    let mt = Math.trunc((btrd - dt) * 60);
                    let st = (btrd - dt - mt / 60) * 3600;
                    console.log("发信点对收信点的方位角为" + dt + "度" + mt + "分" + st + "秒");
                    let dr = Math.trunc(brtd);
                    let mr = Math.trunc((brtd - dr) * 60);
                    let sr = (brtd - dr - mr / 60) * 3600;
                    console.log("收信点对发信点的方位角为" + dr + "度" + mr + "分" + sr + "秒");
                    process.exit(0);
                });
            });
        });
    });
}

main()